### ISSUE TITLE
Unable to create router and nat gateway.

### ISSUE THREAD

**Issue by ReenaAcharya** [opentelekomcloud/terraform-provider-opentelekomcloud#127]
Hi,

I am trying to create networking components using terraform template. I have created VPC and subnets. I need to create Load Balancer, router and Nat gateway to have Internet access. In the router, I saw the component as "external gateway" but I didnt find any resource which can be used to create external gateway.
I saw the snippet for NAT gateway, it requires router_id and snippet for router requires external gateway. I don't know how to create or get details for external gateway.

Any suggestion here would be very helpful.

Thanks,
Reena

**Comment by niuzhenguo** [comment]
@ReenaAcharya Sorry for the late reply. In fact you can use VPC directly instead of router, you can bind EIP to Load Balncer port and also you can add a snat rule to NAT gateway to have Internet access.

**Comment by thomaswetzler** [comment]
Hi @all,

to be honest, I do not quite understand the solution. Sorry, but I'm quite new to terraform. 

I create my network with that code:

# Create VPC
resource "opentelekomcloud_vpc_v1" "vpc" {
  name = "${var.vpc_name}"
  cidr = "${var.vpc_cidr}"
}

# Create Subnets
resource "opentelekomcloud_vpc_subnet_v1" "vpc-network" {
  count      = "${length(var.subnets)}"
  name       = "${element(var.subnets, count.index)}"
  vpc_id     = "${opentelekomcloud_vpc_v1.vpc.id}"
  cidr       = "${cidrsubnet(var.vpc_cidr, 8, format("%d0", count.index+1))}"
  gateway_ip = "${cidrhost(cidrsubnet(var.vpc_cidr, 8, format("%d0", count.index+1)), 1)}"
  primary_dns = "100.125.4.25"
  secondary_dns = "8.8.4.4"
}

A router is generated automatically. Within the UI, I can enable the snat option with one click. Within the description of the two resources I couldn't find any parameter for that. Can anyone paste a code snip how to enable snat?

Thanks in advance 

   Thomas

**Comment by ghost** [comment]
> A router is generated automatically. Within the UI, I can enable the snat option with one click. Within the description of the two resources I couldn't find any parameter for that. Can anyone paste a code snip how to enable snat?
> 
> Thanks in advance
> 
> Thomas

Hello Thomas,

I just found the solution for this problem some minutes ago after searching for hours. To create an SNAT just fill the following snippet with the appropriate IDs.
```
resource "opentelekomcloud_nat_gateway_v2" "nat_gateway" {
  internal_network_id = "<ID of the subnet the NAT gateway should be attached to>"
  name = "nat_gateway" // or whatever you want
  router_id = "${opentelekomcloud_vpc_v1.vpc.id}"
  spec = "1" // 1 (small) to 4 (largest)
}
```
Then create an EIP
```
resource "opentelekomcloud_vpc_eip_v1" "nat_eip" {
  "bandwidth" {
    name = "nat_gateway_bandwidth"
    share_type = "PER"
    size = 1000
  }
  "publicip" {
    type = "5_bgp"
  }
}
```
And one or more SNAT rules
```
resource "opentelekomcloud_nat_snat_rule_v2" "snat_rules" {
  floating_ip_id = "${opentelekomcloud_vpc_eip_v1.nat_eip.id}"
  nat_gateway_id = "${opentelekomcloud_nat_gateway_v2.nat_gateway.id}"
  network_id = "<ID of the subnet that needs access to the internet>"
}
```

Hope this helps.

Holger

**Comment by thomaswetzler** [comment]
Hi Holger,

thanks for your explanation. Unfortunately I was looking for a possibility to switch the "SNAT" Tag within the VPC Creation. I don't like to set up a standalone NAT Gateway recommended from you for two reasons:
   - Cost (Additional VM and EIP)
   - Security 

Enabling the SNAT should be possible. It is described here how to do it with the OTC Client (https://cloud.telekom.de/de/infrastruktur/open-telekom-cloud/tutorials/snat-in-der-open-telekom-cloud - be carefull: German! :-)) It should be possible, because it was with the "old" interface too. 
You even could control, if the Flag was set by a simple query:

data "opentelekomcloud_vpc_v1" "vpc" {
   name = "VPC-PaaS"
}
output "vpc_shared" {
   value = "${data.opentelekomcloud_vpc_v1.vpc.shared}"
}

But it seems to me, I'm to blind to find the right attribute to switch the shard flag within terraform. 

Can anyone help?

Thanks & BR

   Thomas

**Comment by ghost** [comment]
Sorry for my misunderstanding. I thought you were talking about the NAT gateway. It looks like it is not possible to set the option by terraform for now.

Scrolling through the code it seems neither this provider nor the huaweicloud SDK which is used under the hood is able to set anyhing beside name and cidr.

UPDATE:
Maybe you should take a look at the examples folder. It seems the whole examples aren't using the vpc but create the network manually and the enable_snat can be set on router.

**Comment by niuzhenguo** [comment]
@thomaswetzler I don't find the option on the API doc here https://docs.otc.t-systems.com/en-us/api/vpc/en-us_topic_0020090608.html
But you can try to use networking_router resource instead which has a enable_snat option https://www.terraform.io/docs/providers/opentelekomcloud/r/networking_router_v2.html#enable_snat

**Comment by robert1x2** [comment]
@thomaswetzler - did recommendation help?

**Comment by niuzhenguo** [comment]
@thomaswetzler `shared` option is now available for VPC resource with #333

**Comment by thomaswetzler** [comment]
Thanks. That works:

# Create VPC
resource "opentelekomcloud_vpc_v1" "vpc" {
  name = "${var.vpc_name}"
  cidr = "${var.vpc_cidr}"
  **shared = true**
}

BR
